src/Microsoft.AspNetCore.ResponseCompression/BodyWrapperStream.cs

```diff
@@ -0,0 +1,185 @@
+            }
+
+            set
+            {
+                throw new NotSupportedException();
+            }
+        }
+
+        public override void Flush()
+        {
+            if (UncompressedStream != null)
+            {
+                UncompressedStream.Flush();
+            }
+            else
+            {
+                _bodyOriginalStream.Flush();
+            }
+        }
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            if (UncompressedStream != null)
+            {
+                return UncompressedStream.FlushAsync(cancellationToken);
+            }
+            return _bodyOriginalStream.FlushAsync(cancellationToken);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            OnWrite();
+
+            if (UncompressedStream != null)
+            {
+                UncompressedStream.Write(buffer, offset, count);
+            }
+            else
+            {
+                _bodyOriginalStream.Write(buffer, offset, count);
+            }
+        }
+
+#if NET451
+        public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback callback, object state)
+        {
+            OnWrite();
+
+            if (UncompressedStream != null)
+            {
+                return UncompressedStream.BeginWrite(buffer, offset, count, callback, state);
+            }
+            return _bodyOriginalStream.BeginWrite(buffer, offset, count, callback, state);
+        }
+
+        public override void EndWrite(IAsyncResult asyncResult)
+        {
+            OnWrite();
+
+            if (UncompressedStream != null)
+            {
+                UncompressedStream.EndWrite(asyncResult);
+            }
+            else
+            {
+                _bodyOriginalStream.EndWrite(asyncResult);
+            }
+        }
+#endif
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            OnWrite();
+
+            if (UncompressedStream != null)
+            {
+                return UncompressedStream.WriteAsync(buffer, offset, count, cancellationToken);
+            }
+            return _bodyOriginalStream.WriteAsync(buffer, offset, count, cancellationToken);
+        }
+
+        private void OnWrite()
+        {
+            if (!_compressionChecked)
+            {
+                if (IsCompressable())
+                {
+                    UncompressedStream = new MemoryStream();
```

Comment:
Better, at least now you only buffer responses that are compressible. You can take it one step further though. Instead of making a memory stream here you could directly create the compression stream over the original response stream and update the headers immediately. You'll loose your _minimumSize feature (unless you check the to see if the content-length is already set), and you'll have to use chunked rather than content-length, but that should be OK.